Move godeps to vendor/

thockin opened the pull request:
This is a first-step towards glide support, maybe we don't want or need to take this, but it was easy to try.

This fails to compile, not sure why:

```
# k8s.io/kubernetes/pkg/apis/extensions/v1beta1
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:2703: undefined: extensions.ClusterAutoscaler
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:2703: undefined: ClusterAutoscaler
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:2719: undefined: extensions.ClusterAutoscaler
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:2719: undefined: ClusterAutoscaler
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:2723: undefined: extensions.ClusterAutoscalerList
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:2723: undefined: ClusterAutoscalerList
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:3468: Convert_extensions_JobSpec_To_v1beta1_JobSpec redeclared in this block
    previous declaration at _output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion.go:328
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:3845: Convert_extensions_ScaleStatus_To_v1beta1_ScaleStatus redeclared in this block
    previous declaration at _output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion.go:98
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:4737: Convert_v1beta1_JobSpec_To_extensions_JobSpec redeclared in this block
    previous declaration at _output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion.go:380
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:5186: Convert_v1beta1_ScaleStatus_To_extensions_ScaleStatus redeclared in this block
    previous declaration at _output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion.go:120
_output/local/go/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/conversion_generated.go:2723: too many errors
!!! Error in /home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/hack/lib/golang.sh:417
```

thockin committed:
Convert everything to use vendor/

thockin commented:
@krousey

krousey commented on hack/verify-govet.sh:
```diff
@@ -58,7 +58,7 @@ for i in $targets
     # The intended result is that each incantation of this line returns
     # either 0 (pass) or 1 (fail).
-    godep go vet "${goflags[@]:+${goflags[@]}}" "$i" 2>&1 \
+    go vet "${goflags[@]:+${goflags[@]}}" "$i" 2>&1 \
```
On line 42 above, you might want to add something to filter out vendor since `go list` will list vendored packages (unlike Godeps/_workspace).

thockin commented:
Rebased.  'go build' works but Make still fails.

thockin commented:
I know why.  It's because we're trying to build ginkgo, but unlike _workspace/src/, vendor/ is not considered for build.  Easy fix, testing now.

thockin commented:
One reference to _workspace is left:  hack/update-godep-licenses.sh

thockin commented:
Uggh.  We have deps that do import-rewriting.  We `godep save`ing them, godep does no un-rewrite them.  Until we update all those deps, this PR is hung.

glide, conveniently, DOES un-rewrite them.

```
$ find vendor/ -name _workspace
vendor/k8s.io/heapster/Godeps/_workspace
vendor/github.com/opencontainers/runc/Godeps/_workspace
vendor/github.com/appc/spec/Godeps/_workspace
vendor/github.com/appc/cni/Godeps/_workspace
vendor/github.com/coreos/rkt/Godeps/_workspace
vendor/github.com/google/cadvisor/Godeps/_workspace
vendor/github.com/mesos/mesos-go/examples/Godeps/_workspace
```

thockin commented:
I can't, for the life of me, figure out how to get godep to update `github.com/appc/spec/schema`, which is needed to fix the rewriting mess,.

```
$ go get -u github.com/appc/spec/schema
# cd /home/thockin/tmp/godep-vendor/src/github.com/appc/spec; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/appc/spec/schema: exit status 1
```

hmm, ok...

```
$ rm -rf $GOPATH/src/github.com/appc/spec/
$ go get github.com/appc/spec/schema
$ godep update github.com/appc/spec/schemagodep: Package (github.com/appc/spec/Godeps/_workspace/src/github.com/camlistore/go4/errorutil) not found
$ rm -rf vendor/github.com/appc/spec/schema/
$ godep update github.com/appc/spec/schema
godep: Package (github.com/appc/spec/Godeps/_workspace/src/github.com/camlistore/go4/errorutil) not found
$ vi Godeps/Godeps.json # remove camlistore/go4
$ godep update github.com/appc/spec/schema
godep: no packages can be updated
```

Yet again, godep is just inscrutable.  I have no idea what it wants.

bprashanth commented:
cd /home/thockin/tmp/godep-vendor/src/github.com/appc/spec; git checkout master
godep restore

usually works

thockin commented:
```
$ vi Godeps/Godeps.json # remove camlistore/go4
$ godep update github.com/appc/spec/...
godep: Package (github.com/tobi/airbrake-go) not found
```

This is referenced from logrus in some code we don't use.  If I `go get` that dep and the retry, it spins for 30 seconds and gives me another.  And another.  And another.  Good grief, this is a pain in the butt.  What am I doing wrong?

thockin commented:
Ahh, and it ends in failure:

```
$ godep update github.com/appc/spec/...
godep: Package (code.google.com/p/go-charset/charset) not found
$ go get code.google.com/p/go-charset/charset
package code.google.com/p/go-charset/charset: unable to detect version control system for code.google.com/ path
```

thockin commented:
```
# cd /home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api/resource;
git checkout e310e619fc1ac4f3238bf5ebe9e7033bf5d47ee2
error: Your local changes to the following files would be overwritten by
checkout:
Godeps/Godeps.json
vendor/github.com/appc/spec/schema/image.go
vendor/github.com/appc/spec/schema/pod.go
vendor/github.com/appc/spec/schema/types/isolator_resources.go
vendor/github.com/appc/spec/schema/types/semver.go
Please, commit your changes or stash them before you can switch branches.
Aborting
godep: error restoring dep (k8s.io/kubernetes/pkg/api/resource): exit
status 1
godep: Error restoring some deps. Aborting check.
```

That is trying to restore a specific version of kubernetes.

On Tue, Apr 19, 2016 at 10:21 PM, Prashanth B <notifications@github.com>
wrote:

> cd /home/thockin/tmp/godep-vendor/src/github.com/appc/spec; git checkout
> master
> godep restore
>
> usually works
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-212263392>
>

thockin commented:
Awesome, now Go just barfs.

```
$ go get golang.org/x/net/trace
# cd /home/thockin/tmp/godep-vendor/src/golang.org/x/net; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package golang.org/x/net/trace: exit status 1
```

Nuke from orbit and start over.  Have I mentioned today how much I loathe these tools.  Please just tell me WTF you are doing...please...

thockin commented:
Pushed another try, this time updating appc/spec to current at the same time.  The tools are ALL SORTS of broken here.

thockin commented:
added a fix to jenkins test to `go install ./cmd/...` rather than `./...`, which includes vendor (why?!?)

krousey commented:
https://github.com/golang/go/issues/11659

thockin commented:
@wojtek-t help?

verify-codecgen is failing - down the call tree, codecgen calls `go run` on a generated file.  This is failing to find the ugorji dependency, which is in the vendor directory now (rather than GOPATH).  Do you have any clues?

thockin commented:
More weird behavior...

```thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ GOPATH=~/tmp/godep-vendor

thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ go run codecgen-main-1234.generated.go
helpers.go:36:2: cannot find package "_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/github.com/davecgh/go-spew/spew" in any of:
	/usr/local/go/src/_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/github.com/davecgh/go-spew/spew (from $GOROOT)
	/home/thockin/tmp/godep-vendor/src/_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/github.com/davecgh/go-spew/spew (from $GOPATH)
codecgen-pkg-1234.generated.go:5:2: cannot find package "_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/github.com/ugorji/go/codec" in any of:
	/usr/local/go/src/_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/github.com/ugorji/go/codec (from $GOROOT)
	/home/thockin/tmp/godep-vendor/src/_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/github.com/ugorji/go/codec (from $GOPATH)
context.go:23:2: cannot find package "_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/golang.org/x/net/context" in any of:
	/usr/local/go/src/_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/golang.org/x/net/context (from $GOROOT)
	/home/thockin/tmp/godep-vendor/src/_/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/vendor/golang.org/x/net/context (from $GOPATH)
```

I have no idea what Go is doing with the _ here...

krousey commented:
@thockin can you `go list` in the pkg/api dir? what is codecgen-main-1234.generated.go? Can you put that in a gist?

thockin commented:
@krousey 

```
thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ go list
k8s.io/kubernetes/pkg/api
```

```
thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ cat codecgen-main-1234.generated.go 
//+build ignore

package main
import "."
func main() {
	api.CodecGenTempWrite1234()
}
```

thockin commented:
```
thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ head -11 codecgen-pkg-1234.generated.go

package api

import (
	codec1978 "github.com/ugorji/go/codec"
	"os"
	"reflect"
	"bytes"
	"strings"
	"go/format"
)
```

thockin commented:
For context, I am trying to manually do what verify-codecgen is doing...

thockin commented:
```
thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ go build .
thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ 
```

It builds.  It's just `go run` that is messing up...

krousey commented:
Something is wrong with your `$GOPATH`. That's the only reason it'd search that many places for a package. Is that error message coming from the script? Or after you manually set up your GOPATH?

krousey commented:
https://github.com/kubernetes/kubernetes/blob/f6e5cbe3416f0f0e3826ac3cf6f4ec0f9f8c70d0/hack/lib/golang.sh#L265 does some weird path manipulation that may not be valid with vendoring.

thockin commented:
Yes, it is the GOPATH.  Part of our build sets up a temporary GOPATH and uses it, which horks this totally

```
thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ echo $GOPATH
/home/thockin/tmp/godep-vendor

thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ /tmp/codecgen.T0xj44/codecgen_binary -d 1234 -o types.generated.go.1tmp types.go

thockin@thockin-glaptop2:~/tmp/godep-vendor/src/k8s.io/kubernetes/pkg/api$ GOPATH=/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/_output/local/go /tmp/codecgen.T0xj44/codecgen_binary -d 1234 -o types.generated.go.1tmp types.go
codecgen error: error running 'go run codecgen-main-1234.generated.go': exit status 1, console: helpers.go:36:2: cannot find package "github.com/davecgh/go-spew/spew" in any of:
	/usr/local/go/src/github.com/davecgh/go-spew/spew (from $GOROOT)
	/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/_output/local/go/src/github.com/davecgh/go-spew/spew (from $GOPATH)
codecgen-pkg-1234.generated.go:5:2: cannot find package "github.com/ugorji/go/codec" in any of:
	/usr/local/go/src/github.com/ugorji/go/codec (from $GOROOT)
	/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/_output/local/go/src/github.com/ugorji/go/codec (from $GOPATH)
context.go:23:2: cannot find package "golang.org/x/net/context" in any of:
	/usr/local/go/src/golang.org/x/net/context (from $GOROOT)
	/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/_output/local/go/src/golang.org/x/net/context (from $GOPATH)
foobar.go:21:2: cannot find package "speter.net/go/exp/math/dec/inf" in any of:
	/usr/local/go/src/speter.net/go/exp/math/dec/inf (from $GOROOT)
	/home/thockin/tmp/godep-vendor/src/k8s.io/kubernetes/_output/local/go/src/speter.net/go/exp/math/dec/inf (from $GOPATH)
```

Maybe we've finally reached the time where you just HAVE TO use Go's directory structure...

krousey commented:
Is there a reason not to?

thockin commented:
It's sort of a breaking change...sort of..

Maybe `cd $(pwd -P | sed "s|${KUBE_ROOT}|${KUBE_GOPATH}/src/${KUBE_GO_PACKAGE}|")` - it gets farther but fails the same way in another dir....

thockin commented:
removing the crazy GOPATH trickery *does* seem to make this pass...

thockin commented:
It means that anyone who symlinks kube -> gopath/src/k8s.io/kubernetes will no longer be able to build.  I do that.  I think @zmerlynn does that.  I can live with it.

krousey commented:
So what needed that trickery? We still get the _output directory with `go build -o` right?

krousey commented:
The cross build image?

thockin commented:
I'll have to analyze the impact of changing this...sigh...  This is pain we brought on ourselves, I guess

zmerlynn commented:
I don't think you'll break me. If you break me, I'll live. :) Just provide good docs on what the right workflow is, in case we don't all breathe the normal `go` way.

thockin commented:
GRGAGGAGHAHH!  It worked once and now won't work again.   I am walking away for a while.

thockin commented:
Nailed it.  It's because "testdata" is a special directory name and therefore isn't (something something mumble) by the tools.  Can we rename `cmd/libs/go2idl/client-gen/testdata` or will that have deleterious effects?

krousey commented:
@thockin so this is the urgorogirgrgijrigjroijorijgoirj codec puking on the client-gen test data? Or is it the actual client-gen breaking? If it's just urgorjigorijgoirgjiroijgroigjroijgroijorijgr, I see no reason that the codec has to be generated for this test data. @caesarxuchao What do you think?

wojtek-t commented:
>
verify-codecgen is failing - down the call tree, codecgen calls go run on a generated file. This is failing to find the ugorji dependency, which is in the vendor directory now (rather than GOPATH). Do you have any clues?

Not really (I've always had problems with godeps).

thockin commented:
ugorji is being invoked to ensure that the testdata `types.go` did not
change, which would necessitate regenerating code.  Maybe we could just
not?  I don't know this area well..

@wojtek-t what would happen if I intentionally filtered testdata from the
verify/udpate codecgen scripts?

On Wed, Apr 20, 2016 at 11:17 PM, krousey <notifications@github.com> wrote:

> @thockin <https://github.com/thockin> so this is the
> urgorogirgrgijrigjroijorijgoirj codec puking on the client-gen test data?
> Or is it the actual client-gen breaking? If it's just
> urgorjigorijgoirgjiroijgroigjroijgroijorijgr, I see no reason that the
> codec has to be generated for this test data. @caesarxuchao
> <https://github.com/caesarxuchao> What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-212762513>
>

wojtek-t commented:
@thockin - those codecgen are just used for performance improvements. So not running it for test types is not a big deal.

thockin commented:
if I filter */testdata/* it passes that test...

On Thu, Apr 21, 2016 at 12:38 AM, Tim Hockin <thockin@google.com> wrote:

> ugorji is being invoked to ensure that the testdata `types.go` did not
> change, which would necessitate regenerating code.  Maybe we could just
> not?  I don't know this area well..
>
> @wojtek-t what would happen if I intentionally filtered testdata from the
> verify/udpate codecgen scripts?
>
> On Wed, Apr 20, 2016 at 11:17 PM, krousey <notifications@github.com>
> wrote:
>
>> @thockin <https://github.com/thockin> so this is the
>> urgorogirgrgijrigjroijorijgoirj codec puking on the client-gen test data?
>> Or is it the actual client-gen breaking? If it's just
>> urgorjigorijgoirgjiroijgroigjroijgroijorijgr, I see no reason that the
>> codec has to be generated for this test data. @caesarxuchao
>> <https://github.com/caesarxuchao> What do you think?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-212762513>
>>
>
>

wojtek-t commented:
@thockin - can this be related to #24473 ?

caesarxuchao commented:
The testdata does need the codec, so filtering that path out is not an option.

thockin commented:
We can keep the generated code we have - the question is whether we ever
need to re-codecgen it?

On Thu, Apr 21, 2016 at 10:08 AM, Chao Xu <notifications@github.com> wrote:

> The testdata does need the codec, so filtering that path out is not an
> option.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-213020230>
>

caesarxuchao commented:
I can't promise I won't add new fields/types to that test.

The codec doesn't need to be generated/verified by the script though. Can you still manually run ugorji?

thockin commented:
no.  It seems that `go run` will not find deps when run from within
`testdata`, at least that is my best guess at WTF go is doing

On Thu, Apr 21, 2016 at 10:29 AM, Chao Xu <notifications@github.com> wrote:

> I can't promise I won't add new fields/types to that test.
>
> The codec doesn't need to be generated/verified by the script though. Can
> you still manually run ugorji?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-213028084>
>

krousey commented:
Does `go run -x` give any more information?

thockin commented:
nope

On Thu, Apr 21, 2016 at 10:43 AM, krousey <notifications@github.com> wrote:

> Does go run -x give any more information?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-213033176>
>

thockin commented:
Worse now, I have 2 tabs open - identical in every way that I can see.  One
of them passes (once I filter out testdata) and one does not.  I have no
idea what is different


On Thu, Apr 21, 2016 at 10:49 AM, Tim Hockin <thockin@google.com> wrote:

> nope
>
> On Thu, Apr 21, 2016 at 10:43 AM, krousey <notifications@github.com>
> wrote:
>
>> Does go run -x give any more information?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-213033176>
>>
>
>

thockin commented:
The difference between tabs: GO15VENDOREXPERIMENT=1

thockin commented:
ugorji actually looks at that variable.  SONOFA....

eparis commented:
WTH?

krousey commented:
cc @spxtr

thockin commented:
OK, I think I have it building, as long as the testdata ugorji target is filtered.  A few things left to do..

thockin commented:
also need to rebase, still

thockin committed:
Godeps licenses

Fix the licenses script and run it.

thockin commented:
Rebased.  Tests pass locally.  verify-codecgen passes.  This is the part where it fails on Travis.  Ten bucks....

thockin committed:
Demand at least go1.6

thockin commented:
Oh, FFS rebase already?

thockin committed:
Fix path munging funcs and usage

Our `realpath` and `readlink -f` functions (required only because of MacOS,
thanks Steve) were poor substitutes at best.  Mostly they were downright
broken.  This thoroughly overhauls them and adds a test (in comments, since we
don't seem to have shell tests).  For all the interesting cases I could think
of, the fakes act just like the real thing.

Then use those and canonicalize KUBE_ROOT.  In order to make recursive calls of
our shell tool not additively grow `pwd` we have to essentially make the
sourcing of init.sh idempotent.

thockin commented:
OK, I think I have fixed the path-related mess.  Rebasing AGAIN and will see what CI says about it...

thockin commented:
@lavalamp @wojtek-t it looks like something in go2idl doesn't understand the vendor/ dir, and is therefore not finding packages.

```
cmd/libs/go2idl/parser/parse.go:34:2: cannot find package "github.com/golang/glog" in any of:
	/usr/local/go/src/github.com/golang/glog (from $GOROOT)
	/go/src/k8s.io/kubernetes/_output/local/go/src/github.com/golang/glog (from $GOPATH)
cmd/libs/go2idl/go-to-protobuf/main.go:24:2: cannot find package "github.com/spf13/pflag" in any of:
	/usr/local/go/src/github.com/spf13/pflag (from $GOROOT)
	/go/src/k8s.io/kubernetes/_output/local/go/src/github.com/spf13/pflag (from $GOPATH)
cmd/libs/go2idl/go-to-protobuf/protoc-gen-gogo/main.go:25:2: cannot find package "github.com/gogo/protobuf/gogoproto" in any of:
	/usr/local/go/src/github.com/gogo/protobuf/gogoproto (from $GOROOT)
	/go/src/k8s.io/kubernetes/_output/local/go/src/github.com/gogo/protobuf/gogoproto (from $GOPATH)
cmd/libs/go2idl/go-to-protobuf/protoc-gen-gogo/main.go:26:2: cannot find package "github.com/gogo/protobuf/proto" in any of:
	/usr/local/go/src/github.com/gogo/protobuf/proto (from $GOROOT)
	/go/src/k8s.io/kubernetes/_output/local/go/src/github.com/gogo/protobuf/proto (from $GOPATH)
cmd/libs/go2idl/go-to-protobuf/protoc-gen-gogo/main.go:27:2: cannot find package "github.com/gogo/protobuf/sortkeys" in any of:
	/usr/local/go/src/github.com/gogo/protobuf/sortkeys (from $GOROOT)
	/go/src/k8s.io/kubernetes/_output/local/go/src/github.com/gogo/protobuf/sortkeys (from $GOPATH)
cmd/libs/go2idl/go-to-protobuf/protoc-gen-gogo/main.go:22:2: cannot find package "github.com/gogo/protobuf/vanity/command" in any of:
	/usr/local/go/src/github.com/gogo/protobuf/vanity/command (from $GOROOT)
	/go/src/k8s.io/kubernetes/_output/local/go/src/github.com/gogo/protobuf/vanity/command (from $GOPATH)
```

I can dig in but you might get there faster.

thockin committed:
update third_party go/build to go1.6

thockin commented:
@lavalamp @wojtek-t It looks like updating third_party go/build brings some support for vendor dirs, BUT it doesn't work.  I am pretty sure it doesn't work because we pass "." for srcDir (b.context.Import(), as in cmd/libs/go2idl/parser/parse.go:606.

This seems like a more significant overhaul than I was hoping for, since those callsites don't seem to have a "current dir" unless it's buried somewhere in build.Context?

thockin commented:
I won't have much time this week, but this is soundly in my way now.  Help?

thockin commented:
If I pass `os.Getwd()` instead of ".", it gets farther, but now blows up with 

```
type checker error: k8s.io/kubernetes/pkg/api/context.go:23:2: could not import golang.org/x/net/context (No files for pkg "golang.org/x/net/context": 
```

followed by an enormous (hundreds of pages?) struct dump.

thockin commented:
Hmm, maybe fix isn't so hard.  It is not strictly generically correct, but good enough for us.

thockin committed:
Make go2idl handle vendor dirs

thockin commented:
@lavalamp @wojtek-t PTAL at last commit re: go2idl

lavalamp commented on cmd/libs/go2idl/parser/parse.go:
```diff
@@ -127,21 +143,30 @@ func (b *Builder) buildPackage(pkgPath string) (*build.Package, error) {
-	return b.addFile(name, src, true)
+// AddFile adds a file to the set. The pkg must be of the form
+// "canonical/pkg/path" and the path must be the absolute path to the file. A
+func (b *Builder) AddFile(pkg string, name string, src []byte) error {
```
s/name/path/

thockin replied:
done

On Mon, Apr 25, 2016 at 1:10 PM, Daniel Smith <notifications@github.com>
wrote:

> In cmd/libs/go2idl/parser/parse.go
> <https://github.com/kubernetes/kubernetes/pull/24242#discussion_r60979630>
> :
>
> > @@ -127,21 +143,30 @@ func (b *Builder) buildPackage(pkgPath string) (*build.Package, error) {
> >  	return pkg, nil
> >  }
> >
> > -// AddFile adds a file to the set. The name must be of the form canonical/pkg/path/file.go.
> > -func (b *Builder) AddFile(name string, src []byte) error {
> > -	return b.addFile(name, src, true)
> > +// AddFile adds a file to the set. The pkg must be of the form
> > +// "canonical/pkg/path" and the path must be the absolute path to the file. A
> > +func (b *Builder) AddFile(pkg string, name string, src []byte) error {
>
> s/name/path/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242/files/c45dc75e44061a01d7b49fb7237b572d23b067a3..27902ce681296b1bbf11345fe1c83209fcba49e0#r60979630>
>

lavalamp commented on cmd/libs/go2idl/parser/parse.go:
```diff
@@ -127,21 +143,30 @@ func (b *Builder) buildPackage(pkgPath string) (*build.Package, error) {
-func (b *Builder) AddFile(name string, src []byte) error {
-	return b.addFile(name, src, true)
+// AddFile adds a file to the set. The pkg must be of the form
+// "canonical/pkg/path" and the path must be the absolute path to the file. A
```
s/ A//

thockin replied:
Done

On Mon, Apr 25, 2016 at 1:11 PM, Daniel Smith <notifications@github.com>
wrote:

> In cmd/libs/go2idl/parser/parse.go
> <https://github.com/kubernetes/kubernetes/pull/24242#discussion_r60979701>
> :
>
> > @@ -127,21 +143,30 @@ func (b *Builder) buildPackage(pkgPath string) (*build.Package, error) {
> >  	return pkg, nil
> >  }
> >
> > -// AddFile adds a file to the set. The name must be of the form canonical/pkg/path/file.go.
> > -func (b *Builder) AddFile(name string, src []byte) error {
> > -	return b.addFile(name, src, true)
> > +// AddFile adds a file to the set. The pkg must be of the form
> > +// "canonical/pkg/path" and the path must be the absolute path to the file. A
>
> s/ A//
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242/files/c45dc75e44061a01d7b49fb7237b572d23b067a3..27902ce681296b1bbf11345fe1c83209fcba49e0#r60979701>
>

lavalamp commented:
I think go2idl may eventually need the concept of "starting from X" for its lookups. But for now the working directory is no worse than it was before. I haven't looked at the rest of the PR but the go2idl changes LGTM.

thockin commented:
Thanks, Daniel.  There's more to do - apparently the protobuf generation is
failing now.  Sigh.

On Mon, Apr 25, 2016 at 1:37 PM, Daniel Smith <notifications@github.com>
wrote:

> I think go2idl may eventually need the concept of "starting from X" for
> its lookups. But for now the working directory is no worse than it was
> before. I haven't looked at the rest of the PR but the go2idl changes LGTM.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-214513125>
>

thockin committed:
Make *-generated-protobuf work with vendor

thockin committed:
Get rid of _workspace in jenkins

thockin commented:
@ixdy @fejta  - the jenkins failure mode is pretty inscrutable to me - do you guys have time for an assist?

https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/24242/kubernetes-pull-build-test-e2e-gce/36965/

zmerlynn commented:
```
	chdir ./../../../../../../platforms/linux/amd64: no such file or directory
```

@thockin: Something hinky is going on when it goes to run `ginkgo`. As a first guess, you might need to rummage around the `kube::util::find-binary "ginkgo"` and `kube::util::find-binary "e2e.test"` pieces in `hack/ginkgo-e2e.sh`, but @ixdy might know better.

ixdy commented:
@k8s-bot e2e test this please, issue #IGNORE

I want to try to grab the test tarball from Jenkins.

ixdy commented:
Adding `set -x` to `hack/ginkgo-e2e.sh` would also be helpful for debugging.

ixdy commented:
OK, I downloaded the tarballs, extracted them into `/tmp/jenkins/kubernetes`, and made some progress.

I `cd`d into `/tmp/jenkins/kubernetes`, updated `hack/ginkgo-e2e.sh` to `set -x`, and got the command it was running: `/tmp/jenkins/kubernetes/platforms/linux/amd64/ginkgo /tmp/jenkins/kubernetes/platforms/linux/amd64/e2e.test -- ... a bunch of args`

Just running that seems to work:
```console
$ /tmp/jenkins/kubernetes/platforms/linux/amd64/ginkgo /tmp/jenkins/kubernetes
/platforms/linux/amd64/e2e.test
I0427 18:07:50.033350   10104 e2e.go:57] The --provider flag is not set.  Treating as a conformance test.  Some tests may not be run.
I0427 18:07:50.033787   10104 e2e.go:229] Starting e2e run "a05e23b0-0cdd-11e6-bd2d-aa000100069c" on Ginkgo node 1
Running Suite: Kubernetes e2e suite
===================================
Random Seed: 1461805669 - Will randomize all specs
Will run 212 of 280 specs
```

However, it seems that we've actually moved into a subdirectory by the time `hack/ginkgo-e2e.sh` runs, in particular `/tmp/jenkins/kubernetes/_output/local/go/src/k8s.io/kubernetes`. If I go there first:
```console
$ cd /tmp/jenkins/kubernetes/_output/local/go/src/k8s.io/kubernetes
$ /tmp/jenkins/kubernetes/platforms/linux/amd64/ginkgo /tmp/jenkins/kubernetes/platforms/linux/amd64/e2e.test
Failed to run test suite!
        chdir ./../../../../../../platforms/linux/amd64: no such file or directory
Ginkgo ran 1 suite in 399.447µs
Test Suite Failed
```

I don't have time to investigate what's going wrong here, but I can look more later tonight or tomorrow.

ixdy commented:
Note that `_output/local/go/src/k8s.io` are all regular directories, `_output/local/go/src/k8s.io/kubernetes` is a symlink to `/tmp/jenkins/kubernetes`.

ixdy commented on hack/lib/golang.sh:
```diff
@@ -283,32 +269,53 @@ EOF
+  # upset if this is not true.
+  local subdir
+  subdir=$(kube::realpath . | sed "s|$KUBE_ROOT||")
+  cd "${KUBE_GOPATH}/src/${KUBE_GO_PACKAGE}/${subdir}"
```
ah, this is the bit that's causing problems for some reason.

thockin replied:
We have to do this to make various `go` tools happy with vendor directories - it will not find them if it is not in a cwd that is under the GOPATH.  So now we know this is the root  of the problem, what to do with it?

eparis replied:
It seems we must abandon the idea that kube can be unzipp-ed anywhere and built. We force the user to only download kube into GOPATH and only build in GOPATH. We give up trying to make the kube source self contained. Just hoist the complexity on our users (as golang seems to keep forcing us to do).

We could also try to just pull that out. Wasn't it put in partially for godeps, which we aren't using?

ixdy replied:
Yeah, my inclination is to get rid of the `_output/local/go/src/k8s.io.kubernetes` hack entirely - as @eparis notes, I'm guessing it was added so we could `go build` without using `godep`.

Now that we're abandoning Godeps, `go build` should just work in the root. If it doesn't, we need to fix that.

ixdy replied:
I think our build scripts do far more than they need to nowadays (go 1.2/1.3 were much more limited in functionality, I guess), so we probably have a lot of room available for simplification. That said, you obviously should change as little as possible in this PR.

eparis replied:
It's not just `go build` without godep.  say I download/unzip  the tarball into /home/eparis/kubernetes. I `cd /home/eparis/kubernetes` and type `go build ./...` (or something like that). The first golang import it finds for `k8s.io/kubernetes/pkg/util/` is going to look for that in `$GOPATH/src/k8s.io/kubernetes/pkg/util`.  It won't look in `/home/eparis/kubernetes/pkg/util`.

This is a go thing, not a godep thing.

The sad part if that `vendor/` allows everything except OUR imports to work.

(as I understand it).

spxtr replied:
I think it's perfectly reasonable to require developers to lay out their workspace and set their GOPATH properly.

zmerlynn replied:
Really? There's a bunch of other gorp packaged in the kubernetes tar. What are you planning to do with that?

zmerlynn replied:
(I feel like it doesn't belong in a random GOPATH directory.)

thockin replied:
The symlink _output stuff serves 2 purposes.

1) It allows you to 'make' or 'hack/build-go.sh' in any the kube dir, regardless of where it was untarred (good, IMO) - this means you do NOT have to be in the GOPATH yourself, but it means our scripts *do*.

2) It allows us to collect the final outputs of 'go install' in a known place rather than my own $GOPATH.

Once we have a fully dockerized build, (2) matters less.  I still think (1) is valuable (especially since I believe we will NOT get away from Makefile as our main build interface, but that is a diff topic).  So we *have* to change dirs like this.  We just need to make all our tools understand it.

eparis replied:
Are there and `../` in the path? does it work if you use `cd $(realpath $PATHNAME)`

thockin replied:
There should be no .. in the path at all.  First it does cd $(realpath .),
then changes to ./_output/local/go/src/...

On Thu, Apr 28, 2016 at 3:16 PM, Eric Paris <notifications@github.com>
wrote:

> In hack/lib/golang.sh
> <https://github.com/kubernetes/kubernetes/pull/24242#discussion_r61511614>
> :
>
> >    export GOPATH
> >
> > +  # Change directories so that we are within the GOPATH.  Some tools get really
> > +  # upset if this is not true.
> > +  local subdir
> > +  subdir=$(kube::realpath . | sed "s|$KUBE_ROOT||")
> > +  cd "${KUBE_GOPATH}/src/${KUBE_GO_PACKAGE}/${subdir}"
>
> Are there and ../ in the path? does it work if you use cd $(realpath
> $PATHNAME)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242/files/71a332d7a1bad69f61d141a83280cdf5b804ee77#r61511614>
>

thockin committed:
Fix update-generated-swagger-docs

thockin commented:
I think I have the last bits of unit tests passing.  Still need a clue on e2e :(  This is all WAAYYYYYYY harder than it should be.

thockin committed:
Fix verify-godeps

ixdy commented:
@thockin I suspect something in https://github.com/thockin/kubernetes/blob/godep_vendor_dir/vendor/github.com/onsi/ginkgo/ginkgo/testrunner/test_runner.go#L384-L390 is the problem.

https://github.com/onsi/ginkgo/commit/be259a120bdbe4dd5a3352edac586871aa8589e8 makes me think that if we upgrade Ginkgo,the problem may go away.

There's already a request to upgrade our dep of Ginkgo, too (#24140).

thockin commented:
Rebased again

eparis commented:
looks like the hack/verify-godeps.sh failure is some combination of `.gitignore` and incomplete `git add` in `vendor`   I'll get it run down in the morning.

eparis commented:
Just a heads up on the `.gitignore` v `git add` issue.
`/.gitignore` has `kubernetes`/
When godep copies heapster into vendor it is copying in
`vendor/k8s.io/heapster/Godeps/_workspace/src/k8s.io/kubernetes/BLAH`
But since the .gitignore is not anchored in `/` git happily ignores `vendor/k8s.io/heapster/Godeps/_workspace/src/k8s.io/kubernetes/` and so @thockin didn't `git add` it. When the verification runs the tree in `/tmp/gopath.XXXXXX` has the files in `vendor/k8s.io/heapster/Godeps/_workspace/src/k8s.io/kubernetes/` but the tree does not and boom. So this is easily fixed by anchoring things in `/.gitignore` which should be anchored at `/`.

This still leaves `vendor/github.com/jmespath/go-jmespath/.gitignore` which has `jpgo` unanchored. So godep copies in `vendor/github.com/jmespath/go-jmespath/cmd/jpgo` which in turn gets ignored and thus wasn't added.

@thockin your fastest fix it to
```
godep restore
rm -rf Godeps vendor
godep save ./...
git checkout Godeps
git add vendor/github.com/coreos/rkt/Godeps/_workspace/src/k8s.io/ vendor/k8s.io/heapster/Godeps/_workspace/src/k8s.io/ vendor/github.com/jmespath/go-jmespath/cmd/
```

I'll open a PR to fix our .gitignore issue and I'll see what jmespath says. Its dumb to have Godeps under `vendor` but we'll cross that bridge later...

thockin commented:
nice.  I have another thing I want to try which might simplify some (using GOBIN), I'll do your trick or something on my next rebase (later today, I guess)

thockin commented:
pushed with `git add -f $(git ls-files --other vendor/)`

I have another commit to use GOBIN instead of PATH mucking, but it didn't Just Work, so I backed it out for now.

eparis commented:
```
diff --git a/cmd/libs/go2idl/generator/snippet_writer_test.go b/cmd/libs/go2idl/generator/snippet_writer_test.go
index 19ef83d..3d557c7 100644
--- a/cmd/libs/go2idl/generator/snippet_writer_test.go
+++ b/cmd/libs/go2idl/generator/snippet_writer_test.go
@@ -29,7 +29,7 @@ import (
 func construct(t *testing.T, files map[string]string) *generator.Context {
        b := parser.New()
        for name, src := range files {
-               if err := b.AddFile(name, []byte(src)); err != nil {
+               if err := b.AddFile("/tmp/"+name, name, []byte(src)); err != nil {
                        t.Fatal(err)
                }
        }
diff --git a/hack/test-go.sh b/hack/test-go.sh
index 55a5751..c761ac7 100755
--- a/hack/test-go.sh
+++ b/hack/test-go.sh
@@ -40,6 +40,8 @@ kube::test::find_dirs() {
           -o -path './test/e2e_node/*' \
           -o -path './test/integration/*' \
           -o -path './test/component/scheduler/perf/*' \
+          -o -path './third_party/*'\
+          -o -path './vendor/*'\
         \) -prune \
       \) -name '*_test.go' -print0 | xargs -0n1 dirname | sed 's|^\./||' | sort -u
   )
```

thockin commented:
```
+++ [0503 00:27:49] Running tests without code coverage
# k8s.io/kubernetes/third_party/golang/go/build
package k8s.io/kubernetes/third_party/golang/go/build (test)
	imports internal/testenv: use of internal package not allowed
FAIL	k8s.io/kubernetes/third_party/golang/go/build [setup failed]
```

I'll look at this in a bit unless you get there first

eparis commented:
That patch I posted should fix that error.

thockin commented:
ahh, right!

On Tue, May 3, 2016 at 7:36 AM, Eric Paris <notifications@github.com> wrote:

> That patch I posted should fix that error.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-216548694>
>

eparis commented:
I'm so confused....

eparis commented:
@k8s-bot unit test this issue: #IGNORE

thockin commented:
Rebased again

On Tue, May 3, 2016 at 9:02 AM, Eric Paris <notifications@github.com> wrote:

> @k8s-bot <https://github.com/k8s-bot> unit test this issue: #IGNORE
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-216576956>
>

thockin commented:
@k8s-bot unit test: github issue #IGNORE

thockin committed:
Kill our atomic pkg, now that 1.6 is req'd

thockin committed:
Get rid of hack/after-build scripts

The build is now fast enough to not need them.

thockin commented:
@caesarxuchao Do you have any time to think up a solution to the codecgen vs testdata thing?  It's one of the last issues with this PR

thockin commented:
@ixdy any clues where to look for jenkins issues?

caesarxuchao commented:
Just reading your comment [here](https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-212689002), does renaming the testdata folder help?

Another approach is solving https://github.com/kubernetes/kubernetes/issues/24610, which I don't have time right now.

thockin commented:
@k8s-bot test this github issue: #IGNORE

ixdy commented:
@thockin it may have been lost in the noise, but I think bumping up to the latest `ginkgo` may solve the e2e problems. That'll happen just as soon as the SQ gets around to merging #25081.

ixdy commented:
@k8s-bot e2e test this, issue ##25153

thockin commented:
Ughg, this rebase broke it all...

thockin commented:
@ixdy: jenkins fails in the same way.  Deep inside ginkgo it is creating a relative path and setting an exec.Cmd.Dir field to ./../../../../ and so on...

thockin commented:
Don't mistake this "all green" for success - it's just a modicum of progress.

thockin commented:
Issues left:

1) testdata vs codecgen
2) final ginkgo fix
3) get someone to review it.

@eparis, you could start in on #3 while I beat on the others.

thockin committed:
Rename codecgen "testdata" dir so ugorji works

Go mistreats "testdata" and can't find vendor/ dirs.

thockin commented:
Awesome!  another rebase.

thockin committed:
Don't set the golang env for ginkgo

a) it doesn't need it

b) changing CWD to a path with symlinks breaks deep within ginkgo, where it
crafts a relative path to ../../../../../../platforms/amd64/whatever which then
traverses the physical path not the symlinked one, and breaks.

thockin commented:
OK.  I think all issues are fixed.  Once #25240 goes in I can take the last commit off.

@eparis - back to you for review.

caesarxuchao commented on cmd/libs/go2idl/client-gen/main.go:
```diff
@@ -105,12 +106,12 @@ func main() {
 	if *test {
 		arguments.InputDirs = append(dependencies, []string{
-			"k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testdata/apis/testgroup.k8s.io",
+			"k8s.io/kubernetes/cmd/libs/go2idl/client-gen/test_apis/testgroup.k8s.io",
```
Do you know why ugorji is picky on the folder name?

thockin replied:
"testdata" is a magic word to Go, and any dir named "testdata" is ignored
by the tools.  Apparently this includes not looking at vendor/ dirs.
 ugorji calls `go run` on a file in this dir, thus it fails to compile.

On Thu, May 5, 2016 at 11:20 PM, Chao Xu <notifications@github.com> wrote:

> In cmd/libs/go2idl/client-gen/main.go
> <https://github.com/kubernetes/kubernetes/pull/24242#discussion_r62292673>
> :
>
> > @@ -105,12 +106,12 @@ func main() {
> >
> >  	if *test {
> >  		arguments.InputDirs = append(dependencies, []string{
> > -			"k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testdata/apis/testgroup.k8s.io",
> > +			"k8s.io/kubernetes/cmd/libs/go2idl/client-gen/test_apis/testgroup.k8s.io",
>
> Do you know why ugorji is picky on the folder name?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242/files/9bd7abdc6eff192782c250b5dcfbe486eeabc471..b2a5b40281484bef68f78271d744be7efeae9638#r62292673>
>

caesarxuchao replied:
Cool! Thanks for investigating it.

thockin commented:
Rebased with codecgen fix.

@wojtek-t you might want to look at the go2idl changes in here.  Other than the godeps files themselves, most of the other commits are pretty approachable.

I expect this to pass CI, and I'd REALLLLLY like to get it in ASAP so I can stop rebasing 3 times a day. :)  No pressure folks...

eparis commented:
HOLY COW!   `GCE e2e build/test passed for commit 6e1b5ab.`

eparis commented:
I will review today.

thockin commented:
FFS a rebase already?

eparis commented:
You're still a week from this getting to the top of the merge queue, so I wouldn't rebase toooo often...

spxtr commented:
> You're still a week from this getting to the top of the merge queue, so I wouldn't rebase toooo often...

For a PR this big it's probably fine to put p0 on it to avoid rebase hell.

thockin commented:
yeah, as soon as you say LGTM I am paging oncall to merge this beast.  Every rebase takes a half-hour and requires re-running various tools

thockin commented:
rebased and pushed

thockin commented:
This PR is not as scary as it looks.  PTAL :)

eparis commented on build/common.sh:
```diff
@@ -541,6 +538,11 @@ function kube::build::clean_images() {
 }
 
 function kube::build::ensure_data_container() {
+  # This is temporary, while last remnants of _workspace are obliterated
```
What is this about?  In case there are cached old builds and the cached containers are tainted?

thockin replied:
I'll beef up the comments.

  # This is temporary, while last remnants of _workspace are obliterated.  If
  # the data container exists it might be from before the change from
  # Godeps/_workspace/ to vendor/, and thereby still have a Godeps/_workspace/
  # directory in it, which trips up godep (yay godep!).  Once we are confident
  # that this has run ~everywhere we care about, we can remove this.
  # TODO(thockin): remove this after v1.3 is cut.

eparis commented on hack/lib/golang.sh:
```diff
@@ -293,22 +292,26 @@ EOF
   unset GOBIN
+
+  # This seems to matter to some tools.
+  export GO15VENDOREXPERIMENT=1
```
which tool? even with 1.6?

thockin replied:
godep, ugorji, ginkgo...

yes, even with 1.6

eparis commented on hack/update-codecgen.sh:
```diff
@@ -117,7 +117,7 @@ fi
 echo "Building codecgen"
 CODECGEN="${PWD}/codecgen_binary"
-godep go build -o "${CODECGEN}" github.com/ugorji/go/codec/codecgen
+go build -o "${CODECGEN}" ./vendor/github.com/ugorji/go/codec/codecgen
```
something about the leading `./` seems to bother me, but I can't put my finger on it. Probably sohuld just be quiet.

thockin replied:
leading ./ tells go tools to look for a file relative to CWD.  Without that it looks for $GOPATH/vendor/github.com/ugorji/go/codec/codecgen, which obviously is wrong.

eparis commented on hack/update-godep-licenses.sh:
```diff
@@ -30,6 +30,9 @@ set -o errexit
 set -o pipefail
 
+LANG=C
+LC_ALL=C
```
If this is not exported, it isn't affecting the sort on line 124, is it?  sort without LC_ALL=C == future doom.

thockin replied:
Really?  My whole career I have never once used these...

will export

eparis commented:
I just looked through them all and other than some minor nits it LGTM.

thockin commented:
Nits are addressed and I have to rebase AGAIN.  3rd time today.

thockin commented:
rebased, nits picked, and pushed.

thockin commented:
This obviously won't make it today and I don't expect it to merge over the weekend.   If @eparis marks it LGTM (and nobody else objects), I will rebase it (if needed) and re-apply LGTM on Sunday PM (US time) and maybe we can get WAW oncall to merge it manually?  That way we (US) come it to a working codebase Monday morning.  @wojtek-t

eparis commented:
I'll look it over again and label this weekend.

smarterclayton commented on cmd/libs/go2idl/parser/parse.go:
```diff
@@ -101,8 +107,18 @@ func (b *Builder) AddBuildTags(tags ...string) {
+	// This is a bit of a hack.  The srcDir argument to Import() should
+	// properly be the dir of the file which depends on the package to be
+	// imported, so that vendoring can work properly.  We assume that there is
+	// only one level of vendoring, and that the CWD is inside the GOPATH, so
```
Pretty sure this will break OpenShift and any third party resource groups people create outside of the main kube tree (since they'll have to vendor).  I'll take a look to see what happens.

thockin replied:
I'm assuming this is "FYI" to me.  This is required to make the thing work.

smarterclayton replied:
Reminder to myself - we collapse the vendor hierarchy today, but since more and more items are outside the main repo it'll be a problem for contrib style projects.

smarterclayton commented:
@liggitt fyi for the dependency checker on rebase

thockin committed:
Fixme

thockin committed:
Move deps from _workspace/ to vendor/

godep restore
pushd $GOPATH/src/github.com/appc/spec
git co master
popd
go get go4.org/errorutil
rm -rf Godeps
godep save ./...
git add vendor
git add -f $(git ls-files --other vendor/)
git co -- Godeps/LICENSES Godeps/.license_file_state Godeps/OWNERS

thockin commented:
Rebased and re-pushed.

@k8s-oncall WAW team, consider a manual merge, please?  I rebase this daily :(

thockin commented:
W00t!

eparis commented:
It's hard to know if the merge queue ran really well over the weekend or if this PR just caused every single other PR to need a rebase, so the number in the queue dropped dramtically.  (ok, actually you can tell, this PR seemingly kicked 3 others out of the merge queue because they now need a rebase, so really not that bad)

smarterclayton commented:
The merge queue ran well on Sunday.

On Mon, May 9, 2016 at 11:57 AM, Eric Paris <notifications@github.com>
wrote:

> It's hard to know if the merge queue ran really well over the weekend or
> if this PR just caused every single other PR to need a rebase, so the
> number in the queue dropped dramtically. (ok, actually you can tell, this
> PR seemingly kicked 3 others out of the merge queue because they now need a
> rebase, so really not that bad)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-217906456>
>

MHBauer commented:
wh00! good job!

thockin commented:
FWIW I tried the same treatment on contrib/ (which has many _workspace dirs is various sub-dirs) and godep basically crapped the bed.  First it gave me errors about dirty directories that were not dirty.  Then it decided to vendor the entire contrib directory into a subdir of the contrib directory.  As usual, no error messages, no justification.

bprashanth commented:
I tried doing it for contrib/ingress this morning, I got 
```
package k8s.io/contrib/ingress/controllers/gce
	imports k8s.io/contrib/ingress/controllers/gce/controller
	imports k8s.io/kubernetes/vendor/google.golang.org/api/compute/v1: use of vendored package not allowed
package k8s.io/contrib/ingress/controllers/gce
	imports k8s.io/contrib/ingress/controllers/gce/controller
...
```

At which point I decided to de-prioritize, just rev godeps till right before your commit, and hope someone else gets to the bottom of it :)

smarterclayton commented:
Sounds helpful

eparis commented:
So I tried in contrib. Interesting results. One `godep` issue and now one `golang` issue.

First, if you `rm -rf Godeps` your tree is no longer 'clean' so `godep save` won't run.
Ok, no problem, just `git save` then `godep save ./...`; then `git add --all; git commit -a --amend`

But then I ran into:
```
GOBIN=/storage/gopath/src/k8s.io/contrib/mungegithub CGO_ENABLED=0 GOOS=linux go install -installsuffix cgo -ldflags '-w'
# k8s.io/contrib/mungegithub
./mungegithub.go:109: cannot use "k8s.io/kubernetes/pkg/util/flag".WordSepNormalizeFunc (type func(*"k8s.io/kubernetes/vendor/github.com/spf13/pflag".FlagSet, string) "k8s.io/kubernetes/vendor/github.com/spf13/pflag".NormalizedName) as type func(*"k8s.io/contrib/mungegithub/vendor/github.com/spf13/pflag".FlagSet, string) "k8s.io/contrib/mungegithub/vendor/github.com/spf13/pflag".NormalizedName in argument to root.SetGlobalNormalizationFunc
Makefile:22: recipe for target 'mungegithub' failed
make: *** [mungegithub] Error 2
```

Which looks to me that golang is complaining that `k8s.io/kubernetes/vendor/github.com/spf13/pflag` != `k8s.io/contrib/mungegithub/vendor/github.com/spf13/pflag`

eparis commented:
Found the solution. I had `k8s.io/kubernetes/pkg/util` vendored in but did not have `k8s.io/kubernetes/pkg/util/flag` so `go build` was looking in `$GOPATH/src/k8s.io/pkg/util/flag` instead of in `vendor/k8s.io/pkg/util/flag` and was upset about the results.  Which makes a tee-tiny bit of sense I guess.

corentone commented:
@thockin You mention Glide in the PR and related issue. I was wondering if the choice of keeping Godep versus moving to Glide was documented somewhere?
Maybe the move to vendor was big enough you didn't want to also change the dependency manager? I was also considering both and would definitely love to hear your views on it. Thanks!

thockin commented:
I think glide is a better tool,and intend to move to it as soon as possible.

On Tue, May 10, 2016 at 2:31 PM, Corentin Debains <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> You mention Glide in the PR and
> related issue. I was wondering if the choice of keeping Godep versus moving
> to Glide was documented somewhere?
> Maybe the move to vendor was big enough you didn't want to also change the
> dependency manager? I was also considering both and would definitely love
> to hear your views on it. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24242#issuecomment-218297925>
>
